File: CSAdmin/Clases/SampleIIdentity.vb
Imports System.Data.Objects
Imports System.Data.Objects.DataClasses

Public Class SampleIIdentity
    Implements System.Security.Principal.IIdentity

    Private BDContext As CSAdminBDEntities
    Private nameValue As String
    Private authenticatedValue As Boolean
    Private roleValue As ApplicationServices.BuiltInRole

    Public ReadOnly Property AuthenticationType() As String Implements System.Security.Principal.IIdentity.AuthenticationType
        Get
            Return "Custom Authentication"
        End Get
    End Property

    Public ReadOnly Property IsAuthenticated() As Boolean Implements System.Security.Principal.IIdentity.IsAuthenticated
        Get
            Return authenticatedValue
        End Get
    End Property

    Public ReadOnly Property Name() As String Implements System.Security.Principal.IIdentity.Name
        Get
            Return nameValue
        End Get
    End Property

    Public ReadOnly Property Role() As ApplicationServices.BuiltInRole
        Get
            Return roleValue
        End Get
    End Property

    Public Sub New(ByVal name As String, ByVal pass As String)
        If IsValidNameAndPassword(name, pass) Then
            nameValue = name
            authenticatedValue = True
        Else
            nameValue = ""
            authenticatedValue = False
            roleValue = ApplicationServices.BuiltInRole.Guest
        End If
    End Sub

    Private Function IsValidNameAndPassword( _
        ByVal username As String, _
        ByVal pass As String) As Boolean

        BDContext = New CSAdminBDEntities()
        Dim usuariosQ As ObjectQuery(Of Usuarios) = _
            BDContext.Usuarios.Where("it.Usuario = @usuario AND it.Contra = @pass ")

        usuariosQ.Parameters.Add(New ObjectParameter("usuario", username))
        usuariosQ.Parameters.Add(New ObjectParameter("pass", pass))

        If Not usuariosQ.First.Usuario = "" Then
            Select Case usuariosQ.First.Rol
                Case "ADM"
                    roleValue = ApplicationServices.BuiltInRole.Administrator
                Case "USR"
                    roleValue = ApplicationServices.BuiltInRole.User
                Case Else
                    roleValue = ApplicationServices.BuiltInRole.Guest
            End Select
            Return True
        Else
            Return False
        End If
    End Function
End Class

File: CSAdminApp/Clases/SampleIIdentity.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Security.Principal;
using System.Data.Entity.Core.Objects;
using System.Data.Entity.Core.Objects.DataClasses;

namespace CSAdminApp.Clases
{
    class SampleIIdentity : IIdentity
    {
        #region Campos de IIdentity

        private string nameValue;
        private bool authenticatedValue;
        private WindowsBuiltInRole roleValue;

        #endregion

        #region Propiedades de IIdentity

        public string AuthenticationType
        {
            get { return "Custom Authentication"; }
        }

        public bool IsAuthenticated
        {
            get { return this.authenticatedValue; }
        }

        public string Name
        {
            get { return this.nameValue; }
        }

        public WindowsBuiltInRole Role
        {
            get { return this.roleValue; }
        }

        #endregion

        #region Métodos de IIdentity

        public SampleIIdentity(string name, string pass)
        {
            if (this.IsValidNameAndPassword(name, pass))
            {
                this.nameValue = name;
                this.authenticatedValue = true;
            }
            else
            {
                this.nameValue = "";
                this.authenticatedValue = false;
                this.roleValue = WindowsBuiltInRole.Guest;
            }
        }

        private bool IsValidNameAndPassword(string name, string pass)
        {
            CSAdminBDEntities dataContext = new CSAdminBDEntities();
            var usuariosQ = dataContext.Usuarios
                                       .Where(u => u.Usuario == name && u.Contra == pass );

            if (usuariosQ.Any())
            {
                switch (usuariosQ.First().Rol)
                {
                    case "ADM":
                        roleValue = WindowsBuiltInRole.Administrator;
                        break;

                    case "USR":
                        roleValue = WindowsBuiltInRole.User;
                        break;

                    default:
                        roleValue = WindowsBuiltInRole.Guest;
                        break;
                }
                return true;
            }
            else
            {
                return false;
            }
        }

        #endregion
    }
}
